bftengine/src/bftengine/ReplicaBase.hpp

```diff
@@ -80,11 +81,17 @@ class ReplicaBase {
 
   void sendRaw(MessageBase* m, NodeIdType dest);
 
-  bool validateMessage(MessageBase* msg) {
+  template <typename MessageType>
+  bool validateMessage(MessageType* msg) {
+    if (config_.debugStatisticsEnabled) {
+      DebugStatistics::onReceivedExMessage(msg->type());
+    }
     try {
-      if (config_.debugStatisticsEnabled) DebugStatistics::onReceivedExMessage(msg->type());
-
-      msg->validate(*repsInfo);
+      if constexpr (std::is_same_v<MessageType, CheckpointMsg>) {
+        msg->validate(*repsInfo, false);
```

Comment:
Another question:
We have this mechanism for async validation (see https://github.com/vmware/concord-bft/blob/c769f077da98a4ef6f5ae8fa859d4618c70552d5/bftengine/src/bftengine/ReplicaImp.cpp#L273.
As far as I understand from the code, if async validation is enabled, then the checkpoint message is still being validated. If this is the case, then we need either (1) to fix this for the async validation also or (2) to make sure that checkpoint messages are never async validated (by making sure https://github.com/vmware/concord-bft/blob/c769f077da98a4ef6f5ae8fa859d4618c70552d5/bftengine/src/bftengine/messages/CheckpointMsg.hpp#L73 returns false and asserting on it).
Still, I think I'm missing something, because currently, https://github.com/vmware/concord-bft/blob/c769f077da98a4ef6f5ae8fa859d4618c70552d5/bftengine/src/bftengine/messages/CheckpointMsg.hpp#L73 returns true, and the relevant configuration variable is true by default https://github.com/vmware/concord-bft/blob/c769f077da98a4ef6f5ae8fa859d4618c70552d5/bftengine/include/bftengine/ReplicaConfig.hpp#L209.
Can you clarify?

Reply:
I think that you missed that asyncValidateMessage was already changed to delegate to validateMessage in this branch.
Please resolve this conversation if you are satisfied with this answer.

Reply:
thanks